### DISCUSSION TITLE
Continuation of the discussion #12492

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#15802] [General]
Hi,
I started working on my project again. As a reminder, I want two independent programs to be able to be launched alternately each time a button is pressed..
The demo version works thanks to your help.
I added an Irq trigger to handle the push button
However, I wanted to go further and I'm stuck.
Demo version:
```from machine import Pin
import asyncio
from time import sleep, sleep_ms
flag,go,value = True,False,True

led = Pin(2, Pin.OUT)
led.value(0)
button = Pin(32,Pin.IN,Pin.PULL_DOWN)

def func(v):
  global value,counter
  sleep_ms(50)
  if(button.value() == 0):
    return
  while(button.value() == 1):
    sleep_ms(100)
  sleep_ms(100)
  led.value(value)
  value = 0 if value else 1

button.irq(trigger=Pin.IRQ_RISING, handler=func)

async def prog_1():
    global flag, go
    while True:
        if flag and go:
           print('program 1')
           go = False
        await asyncio.sleep(1)

async def prog_2():
    global flag, go
    while True:
        if not flag and go:
           print('program 2')
           go = False
        await asyncio.sleep(1)

async def main():
    global flag, go
    while True:
        flag = value
        go = True                
        await asyncio.sleep(2)

asyncio.run(asyncio.gather(main(), prog_1(), prog_2()))
```
First question:
in this example, does prog_1 (or prog_2) run for the duration of the main coroutine's asyncio.sleep(2) wait, exit, and start again?
If so, this is not what I want

Second question:
If I run a program with an infinite loop, the button has no effect anymore

coroutine prog_1
```
async def prog_1():
    from program_1 import prog1
    global flag, go
    while True:
        if flag and go:
            prog1()
            go = False
        await asyncio.sleep(1)
``` 
program_1
```
def prog1():
    while True:
        print("prog 1")
    
if __name__ == "__main__":
    prog1()
```
Do you think what I want to do is possible?

**Comment by shariltumin** [comment]
I do not know the details of your project, but if what you want is to mix an interrupt trigger and three async functions, then you might find the following script useful:

```py
import asyncio
from time import sleep, sleep_ms
from machine import Pin, disable_irq, enable_irq
value,state = True,None
go,counter = 0,-1

led = Pin(2, Pin.OUT)
led.value(0)
button = Pin(32,Pin.IN,Pin.PULL_DOWN)

def func(v):
  global value,counter,go,state
  state = disable_irq()
  if go == 0:
     counter+=1
     go=counter%2+1
     value = 0 if value else 1
  enable_irq(state)

button.irq(trigger=Pin.IRQ_RISING, handler=func)

async def prog_1():
    global go, state
    while True:
        if go==1:
           print('program 1')
           go=0
        await asyncio.sleep(1)

async def prog_2():
    global go, state
    while True:
        if go==2:
           print('program 2')
           go=0
        await asyncio.sleep(1)

async def main():
    global counter, value, go
    while True:
        led.value(value)
        if counter>1000000: counter=-1 
        await asyncio.sleep(2)

asyncio.run(asyncio.gather(main(), prog_1(), prog_2()))
```

The running log shows:
```bash
MicroPython v1.23.0 on 2024-06-02; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> import my_p15802
program 1
program 2
program 1
program 2
program 1
program 2
program 1
program 2
program 1
program 2
program 1
...
```

**Reply by PatLub85** [reply]
I put a print after state = disable_irq() and this is what I get:
```
program 1
state; 393248
state; 393248
program 2
Guru Meditation Error: Core  1 panic'ed (Interrupt wdt timeout on CPU1). 
...
Rebooting...
ets Jul 29 2019 12:21:46
...
```
Ouch!

**Reply by shariltumin** [reply]
The IRQ should be as short as possible. It is not a good idea to put a print() in the IRQ. The absolute maximum amount of time that interrupts should be disabled should be less than the timeout of the interrupt watchdog timer. By default, the INT WDT timeout is set to 300ms (or 800ms if SPI RAM is supported). If the INT WDT is triggered, it will only cause a system reset.

**Reply by PatLub85** [reply]
Sorry, I expressed myself badly. I have the same message without the print. I added it to try to understand what was happening.
I also decreased the sleep_ms to 20 in the func function but I have the same error

**Reply by shariltumin** [reply]
You should not have sleep in the IRQ, the ```if go == 0:``` block takes care of the debouncing problem.

**Reply by PatLub85** [reply]
ok now the IRQ works as you recommend.
However I still have my two questions pending

**Reply by shariltumin** [reply]
Question 1: All async tasks are running "at the same time" when one task hits the ```await asyncio.sleep(1)``` it will yield (pause its execution) and hand over execution to the scheduler where the next "ready to run task" will get the CPU time. This will go on forever because all three tasks have ```while True:```

Question 2: You cannot mix async functions with normal functions, because a normal function in your case here "program_1.prog1()" will enter its ```while True:``` and will hold the CPU forever. The IRQ will still happen but your async tasks will never get the a chance to run after that.

**Reply by bixb922** [reply]
I'd use a program similar to the one below. Since program1 and program2 do not need to run simultaneously, only one task is needed, but other asyncio tasks can be added later. No global variables because global variables are sometimes difficult to understand. Calls to program1 and program2 need to be inserted where the print() statements are. Program1 and program2 must return when they have performed their task. They can be async or not.

```py
import asyncio
from machine import Pin

async def wait_for_button_pulsed(b):
	# Wait for button pressed
	while b.value() == 0:
		await asyncio.sleep_ms(50)
	# Wait for button released
	while b.value() == 1:
		await asyncio.sleep_ms(50)
		
async def main():
	led = Pin(2, Pin.OUT)
	led.value(0)
	button = Pin(32,Pin.IN,Pin.PULL_DOWN)
	while True:
		await wait_for_button_pulsed(button)
		print("run program 1 here")
		led.value(1)
		await wait_for_button_pulsed(button)
		print("run program 2 here")
		led.value(0)

asyncio.run(main())
```

**Reply by PatLub85** [reply]
Your program works fine but if I modify it by calling program_1, unfortunately, as shariltumin points out, prog1 enters its while True: and holds the processor forever.
```
async def main():
	button = Pin(32,Pin.IN,Pin.PULL_DOWN)
	while True:
		await wait_for_button_pulsed(button)
		print("run program 1 here")
		from program_1 import prog1
		prog1()
		await wait_for_button_pulsed(button)
		print("run program 2 here")
```
I'm afraid there is no solution to my need.

**Reply by shariltumin** [reply]
If you are using an esp32 based MCU, you may want to consider using the ```\_thread``` library. The ```\_thread``` is a preemptive multitasking system. Each task is preemptively allocated CPU time and no block of code can hold the CPU exclusively. The ```\_thread``` may introduce complexity in your code and subtle bugs that are not easy to debug. But hey, we are here to learn, so why not give it a try.

Or maybe there is an easy way to program a solution for your project using simple sequential control? "Simple is better than complex". Start a regular Python and do ```import this``` at the REPL, you will know what I mean.

**Reply by bixb922** [reply]
program1 and 2 must certainly must return when they have to stop to run.  I still don't follow exactly what you need, perhaps it's something like this?
```py
import asyncio
from machine import Pin

program_number = 0
	
async def program0():
    while True:
        if program_number != 0:
            return
        print("program 0 is running")
        # Program0 and 1 must have a await used periodically in 
        # their while True: loop for this to work.
        await asyncio.sleep(1)

async def program1():
    while True:
        if program_number != 1:
            return
        print("program 1 is running")
        await asyncio.sleep(1)

async def wait_for_button_pulsed(b):
    global program_number
    # Wait for button pressed
    while b.value() == 0:
        await asyncio.sleep_ms(50)
    # Wait for button released
    while b.value() == 1:
        await asyncio.sleep_ms(50)

async def main():
    global program_number
    led = Pin(2, Pin.OUT)
    led.value(0)
    button = Pin(32,Pin.IN,Pin.PULL_DOWN)
    while True:
        await wait_for_button_pulsed(button)
        program_number = 0
        await program0()
        led.value(1)

        await wait_for_button_pulsed(button)
        program_number = 1
        await program1()
        led.value(0)

asyncio.run(main())
```

**Reply by PatLub85** [reply]
I assure you my goal is not to manage a nuclear power plant !
I have a robot car that has two independent programs
With the first the car is autonomous thanks to an ultrasonic detector (hcsr04) and with the second I control it with a microdot server and a web interface on my phone.
My idea is to have the 2 programs in the ESP32 and to switch from one program to the other with a push button.
But more generally, to make several programs coexist and use them alternately? That's my challenge

**Reply by PatLub85** [reply]
For Shariltumin
At the beginning of my reflection I hesitated to use threads or asyncio
I thought that asyncio was simpler
Moreover I used them with Microdot and that's why I tried this track.
I will look at how threads work starting from scratch but as you say, we are here to learn

**Comment by glenn20** [comment]
From my understanding of your use case, I believe it is easily achieved. However, whether you use asyncio or not, your needs require that prog1 and prog2 are written to stop themselves when asked.

Note of clarification:

- In your statement you said:
  - you want prog1 and prog2 "to be launched alternately each time a button is pressed." 
  - But, your program seems to run prog1 and prog2 in parallel and pauses prog1 and resumes prog2, etc.
    - In my proposed solution below, I assume the former interpretation

In either case, you need to write prog1 and prog2 to work cooperatively, eg. so that they will hear and accept a request to stop/pause. If you write an infinite loop into either, they are not working cooperatively. If you want it to work even when either prog behaves uncooperatively, you need a pre-emptive strategy (ie. one that forces a task to end or pause). You **can** use threads to achieve that (but it has traps for the inexperienced and limitations on interacting with hardware from both threads on some platforms).

If your progs can work cooperatively **and** you really do want to **launch** you progs on each button press, a really simple non-asyncio solution can be:

```py
stop_prog: bool = False

def func(v):
  global stop_prog
  stop_prog = True

button.irq(trigger=Pin.IRQ_RISING, handler=func)

while True:
    time.sleep_ms(100) # Switch debounce delay
    stop_prog = False
    prog1()
    time.sleep_ms(100) # Switch debounce delay
    stop_prog = False
    prog2()
```

You need to write your prog1 and prog2 to periodically check the `stop_prog` flag and resond accordingly. eg. change your infinite loop to:
```py
while not stop_prog:
        print("prog 1")
```

> First question: in this example, does prog_1 (or prog_2) run for the duration of the main coroutine's asyncio.sleep(2) wait, exit, and start again? If so, this is not what I want

No. prog1 and prog2 run until they are stopped, but they *may* yield to each other whenever one makes an `await`.

> 
> Second question: If I run a program with an infinite loop, the button has no effect anymore

Yes. If you want to be able to pre-emptively kill or pause a prog unconditionally, you will need something else, eg. threads or just a soft or hard reset. asyncio (and my solution above) rely on cooperation beteween the tasks.

Unless you really have no control over the code for prog1 and/or prog2, it is much simpler just to use the non-asyncio method suggested above.

If you need asyncio for other reasons, you can make that work too, but you need to be careful to ensure `prog2` does not resume execution till `prog1` has agreed to suspend itself. You could use an `asyncio.event` to handle the synchronisation between the tasks (https://docs.micropython.org/en/latest/library/asyncio.html#class-event, https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event)

**Reply by glenn20** [reply]
Oh, and I should also add that if you pursue a pre-emptive strategy, you have to be very careful about interacting with hardware in prog1 and prog2. Killing or pausing a task at the wrong time may leave hardware in an unexpected state when the other prog takes over.

**Reply by PatLub85** [reply]
Hi Glenn,
I tried your solution and my program_1 loops constantly, button.irq has no action and stop_prog never changes
What am I missing?

``` 
while True:
    time.sleep_ms(100) # Switch debounce delay
    stop_prog = False
    from program_1 import prog1
    prog1(stop_prog)
    time.sleep_ms(100) # Switch debounce delay
    stop_prog = False
    print("prog2()")
```
program_1
```
import time
stop_prog=None

def prog1(stop_prog):
    compteur=0
    while not stop_prog:
        compteur +=1
        print("prog 1",compteur)
        time.sleep(1)
    
if __name__ == "__main__":
    prog1(stop_prog)
```

**Reply by mendenm** [reply]
It looks as if stop_prog may not exist in the same namespace since you are importing prog1, and the copy of stop_prog it sees is in the program_1 namespace.

**Reply by PatLub85** [reply]
A priori yes, here are the prints of prog1
```
prog 1 1
stop_prog False
prog 1 2
stop_prog False
prog 1 3
stop_prog False
prog 1 4
stop_prog False
prog 1 5
```

**Reply by PatLub85** [reply]
To make it simpler I inserted the prog1 function in the main program and I have the same result.
The Irq does not switch

**Reply by glenn20** [reply]
Yes, you have created a number of variables called `stop_prog` all running in their own different program scopes (so, ther all seeing different variables called "stop_prog".

(Edit: I mishandled the global scope across modules in my first reply. Should be fixed now.)

If you want to import prog1 and prog2 from separate modules, that makes it a little more complicated, as `global` variables are not shared between modules, but that is easily overcome, but needs a an extra small module to ensure that all files are looking at the same global flag.

Change you code to:


```py
from program_1 import prog1
from program_2 import prog2
import prog_ctrl

def func(v):
  prog_ctrl.stop = True

button.irq(trigger=Pin.IRQ_RISING, handler=func)

while True:
    time.sleep_ms(100) # Switch debounce delay
    prog_ctrl.stop = False
    prog1()
    time.sleep_ms(100) # Switch debounce delay
    prog_ctrl.stop = False
    prog2()
```

prog_crtl.py
```py
# This is the global flag to stop the running program
stop: bool = False
```

program_1:
```py
import time
import prog_ctrl

def prog1():
    compteur=0
    while not prog_ctrl.stop:
        compteur +=1
        print("prog 1",compteur)
        time.sleep(1)

if __name__ == "__main__":
    prog1()
```

**Reply by glenn20** [reply]
And, just in  case you are tempted to change the import statement, you **must** import `prog_ctrl` this way. If you change it like this:

program_1:
```py
from prog_ctrl import stop
```
it will define a new variable called "stop" in your `program_1` module, rather than a reference to the "stop" variable in the `prog_ctrl` module. This is just the way python works.

**Reply by PatLub85** [reply]
It works as I want!!!
Thank you all for your kindness.
As I said above, my second program uses Microdot (so async), I hope I won't have any problem with memory.
By the way, are there any special recommendations with GC?

**Reply by glenn20** [reply]
That's excellent news. 

Re. Gc, it would be a good idea to put `gc.collect()` before each call to `prog1()` and `prog2()`. That will go a long way toward cleaning up the memory used by prog1 before prog2 starts and vice-versa.

**Reply by mendenm** [reply]
In the codes I run with microdot, I do put, in a slow supervisory loop,  an occasional call to `gc.collect()` just to make sure.  This is voodoo programming, but it probably doesn't hurt anything.

**Reply by glenn20** [reply]
If you mean that prog2 calls async.run() (or equivalent) to start microdot as an async task, then that should be fine. You will just need to stop all tasks (including microdot) when `prog_ctrl.stop` is set True before exiting prog2().

**Reply by glenn20** [reply]
If both prog1() and prog2() have large module-scoped memory buffers, they will not be cleaned up by `gc.collect()` and you will just have to have enough RAM to accommodate both (as well as any dynamic memory required by whichever of prog1 and prog2 is running).

**IF** that becomes an issue (it probably won't be), then there are techniques you can use (such as module unloading or rewriting your module to not use large module-scoped buffers or rebooting from prog1 into prog2 using techniques we have discussed before OR buying a module with more RAM 😉).

**Reply by glenn20** [reply]
> If you mean that prog2 calls async.run() (or equivalent) to start microdot as an async task, then that should be fine. You will just need to stop all tasks (including microdot) when `prog_ctrl.stop` is set True before exiting prog2().
> 
> 

Eg. Something like....(Untested)...

```py
from microdot import Microdot 
import prog_ctrl

app = Microdot() 

@app.route('/') 
async def index(request): 
      return 'Hello, world!' 

async def monitor(app):
    while not prog_ctrl.stop:
        await asyncio.sleep_ms(500)
    app.shutdown()

asyncio.run(
    asyncio.gather(
        app.start_server(),
        monitor(app),
    )
)
```

I **believe** `app.shutdown()` will cause the microdot server to start shutting down immediately (and not wait till after the next url access), but if it doesn't, you may need to cancel the task to force it to shutdown.

**Reply by PatLub85** [reply]
Je me suis remis sur mon projet et voici où j'en suis :
L'astuce de Glenn (stop_prog) fonctionne parfaitement.
Par contre j'ai eu des soucis avec les rebond du bouton poussoir.
J'ai trouvé une solution sur ce site https://github.com/jmcclin2/updebouncein/blob/main/README.md
qui donne satisfaction
```
from debounced_input import DebouncedInput
pin = (Pin(18, Pin.IN, Pin.PULL_DOWN))
compteur=0
def callback(pin, pressed,duration_ms ):
    global compteur
    if (pressed):
        compteur +=1
        prog_ctrl.stop = True

button = DebouncedInput(18, callback, pin_pull=Pin.PULL_DOWN)
# and the launch of the two programs
def main():
    while True:
        if compteur % 2 :
            prog_ctrl.stop = False
            gc.collect()
            prog1()
        else:
            prog_ctrl.stop = False
            gc.collect()
            prog2()        

if __name__ == "__main__":
    main()
```
But what I feared happened when I launched the second program that imports Microdot
I have this message: 
```
Traceback (most recent call last):
  File "<stdin>", line 8, in <module>
  File "prog_1_2.py", line 10, in <module>
MemoryError: memory allocation failed, allocating 2344 bytes
```
The line 10
```
from microdot import Microdot,Response,send_file
```
Je vais ouvrir une autre discussion pour ce problème

**Reply by PatLub85** [reply]
Now that I fixed the memory problem, I resumed my project.
As you thought, microdot is problematic. Shutdown does not work properly.
Here is a simple example:
```
from microdot import Microdot
import network
import time
import uasyncio as asyncio
import prog_ctrl

ssid = 'MicroPython-AP'
password = '123456789'
ip_ap='192.168.4.10'
subnet_ap='255.255.255.0'
gateway_ap='192.168.4.1'
dns_ap='192.168.4.1'

def wifi_ap():
    ap = network.WLAN(network.AP_IF)
    ap.active(True)
    ap.ifconfig((ip_ap,subnet_ap,gateway_ap,dns_ap))
    ap.config(essid=ssid, password=password)
    while ap.active() == False:
         pass
    return()
 
wifi_ap()
app = Microdot()

html = '''<!DOCTYPE html>
<html>
    <head>
        <title>Microdot Example Page</title>
        <meta charset="UTF-8">
    </head>
    <body>
        <div>
            <h1>Microdot Example Page</h1>
            <p>Hello from Microdot!</p>
            <p><a href="/shutdown">Click to shutdown the server</a></p>
        </div>
    </body>
</html>
'''

@app.route('/')
async def hello(request):
    return html, 200, {'Content-Type': 'text/html'}

@app.route('/shutdown')
async def shutdown(request):
    prog_ctrl.stop = True

async def monitor(app):
    while not prog_ctrl.stop:
        await asyncio.sleep_ms(500)
    print('stop2',prog_ctrl.stop)
    request.app.shutdown()

def car_web():
    asyncio.run(
        asyncio.gather(app.run(debug=True),monitor(app)))
        
if __name__ == "__main__":
    car_web()
```
By clicking on shutdown, I get : GET /shutdown 204 
and I have the impression that the coro monitor is not activated

**Reply by glenn20** [reply]
Hmmm... in `monitor()`, `request` does not exist in the scope. I **believe** you should be able to just call `app.shutdown()` instead of `request.app.shutdown()`.

**Reply by glenn20** [reply]
Another potential issue is that you should not use `app.run()` from witrhin asyncio.gather() or asyncio.run() (it calls asyncio.run() to start the server. It would be best to use app.start_server() as in my example above.

**Reply by glenn20** [reply]
...and this might explain why the monitor coro is not started.

**Reply by PatLub85** [reply]
Hi Glenn,
your comments are judicious. With this code it works correctly. The push button switches my two programs
```
@app.route('/shutdown')
async def shutdown(request):
    prog_ctrl.stop = True

async def monitor(app):
    while not prog_ctrl.stop:
        await asyncio.sleep_ms(500)
    app.shutdown()

def car_web():
    asyncio.run(
        asyncio.gather(app.start_server(),monitor(app)))
```
The only problem I have left is that when I press the stop button on the web page, the shutdown does not work, while if I use this program alone, it's ok.
I don't know if I'm clear!
Another detail, in a previous version, without the monitor function, when launching the micropython server, before connecting I had the following message: Starting async server on 0.0.0.0:5000...
And now I don't have it anymore

**Reply by glenn20** [reply]
Re. Qn 2, microdot will print that meessage if you set `debug=True` ie: `app.start_server(debug=True)`.

Re. Qn 1, I'm not sure about that, but I suggest trying the following: rename `async def shutdown(request)` to something else (it may clash with some internal microdot functions) and possibly rename '/shutdown' as well - in case microdot treats that route specially.

**Reply by PatLub85** [reply]
Hi,
Qn2 : ok, thanks
Qn1: I made the changes but I have the same result.
However, it seems to me that the callback function of the main program does not increment the counter variable (only when I press the stop button on the web page). Which makes main.py restart prog1
Otherwise, the counter increments normally
main.py
```
from debounced_input import DebouncedInput
compteur=0

def callback(pin, pressed,duration_ms ):
    global compteur
    if (pressed):
        compteur +=1
        prog_ctrl.stop = True

DebouncedInput(18, callback, pin_pull=Pin.PULL_DOWN)

def main():
    while True:
        print("compteur",compteur)
        if compteur % 2 :
            prog_ctrl.stop = False with microdot
            car_web() #prog1
        else:
            prog_ctrl.stop = False
            autonome() #prog2
```
In summary, if I run main.py we have the following sequences:
counter=0 -> prog2
press push button
counter=1 -> prog1
press push button
counter=2 -> prog2
press push button
counter=3 -> prog1
Stop button on the web page
counter=3 -> prog1

**Reply by glenn20** [reply]
Ah, I can see that you have changed the logic in the `main()` function. That would explain the problem.

When you press the button, `proctrl.stop` is set AND the counter is incremented. The microdot `shudown()` route sets `prog_ctrl.stop` but does NOT increment the counter. So, when you press the stop button on the web page, it stops microdot and then starts it up again (because the counter has not been changed.

TO fix, either:

- increment the counter in microdot as well (you will need to make it globally accessible to the microdot code - like we did for `prog_ctrl.stop`) OR
- Change `main()` to:
```py 
def main():
    while True:
        print("compteur",compteur)
        prog_ctrl.stop = False
        car_web() #prog1
        print("compteur",compteur)
        prog_ctrl.stop = False
        autonome() #prog2
```

I would recommend the second approach. The use of the counter is unnecessary for toggling between the programs (and just complicates the business logic, leading to possible errors like this), but it has usefulness as a diagnostic that your button presses are being detected.

**Reply by PatLub85** [reply]
Thinking about it I realize that what is happening is logical. If I am in prog1 (microdot) and I press stop, I stop the server and I return to main.py WITHOUT going through the interruption, therefore without incrementing the counter!
In this logic we can simplify in this way:
main.py
```
def callback(pin, pressed,duration_ms ):
    if (pressed):
        prog_ctrl.compteur = not prog_ctrl.compteur
        prog_ctrl.stop = True
DebouncedInput(18, callback, pin_pull=Pin.PULL_DOWN)
def main():
    while True:
        if prog_ctrl.compteur :
            prog_ctrl.stop = False
            car_web()
        else:
            prog_ctrl.stop = False
            autonome() 
```
and car_web() ->prog1
```
@app.route('/fin')
async def termine(request):
    prog_ctrl.stop = True
    prog_ctrl.compteur = not prog_ctrl.compteur # the modification

async def monitor(app):
    while not prog_ctrl.stop:
        await asyncio.sleep_ms(500)
    app.shutdown()
```
On the other hand, in your second approach I don't understand how the switch between the two programs happens.

**Reply by PatLub85** [reply]
After thinking, I understood your second approach and now everything works
Thanks to all for your help.
With this project I learned to better use async, microdot, javascript.
My only regret is that I didn't get a response to my thread 13529 about fetch.
I didn't get any on javascript forums either!
